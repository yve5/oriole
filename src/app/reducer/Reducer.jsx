import { CVSS4_CHANGE_VALUE, CVSS4_RESET_SCORE } from '../resources/constants';
import { getDefaultMetrics } from '../../cvss4';

const initialState = {
  metrics: getDefaultMetrics(),
};

const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case CVSS4_RESET_SCORE:
      return {
        ...state,
        ...action,
      };

    case CVSS4_CHANGE_VALUE: {
      const updatedMetric = {};
      updatedMetric[action.metric] = action.value;

      const copyAction = { ...action };
      delete copyAction.metric;
      delete copyAction.value;

      return {
        ...state,
        ...copyAction,
        metrics: {
          ...state.metrics,
          ...updatedMetric,
        },
      };
    }

    default:
      return state;
  }
};

export default reducer;
