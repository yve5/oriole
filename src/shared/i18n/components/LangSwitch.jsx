import React from 'react';
import { connect } from 'react-redux';

import I18n from './I18n';
import * as All from '../actions/All';

const LangSwitch = ({ className, switchLang }) => (
  <button
    type="button"
    className={className}
    onClick={() => switchLang()}
    data-testid="i18n-switch-button"
  >
    <I18n>LANGUAGE</I18n>
  </button>
);

const mapStateToProps = (state) => ({
  root: state,
});

const mapDispatchToProps = {
  ...All,
};

export default connect(mapStateToProps, mapDispatchToProps)(LangSwitch);
