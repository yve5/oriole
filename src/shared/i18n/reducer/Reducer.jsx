import en from '../dictionaries/en';
import fr from '../dictionaries/fr';

import { tsl } from '../utils';
import {
  I18N_INIT_LANG,
  I18N_CHANGE_LANG,
  I18N_SWITCH_LANG,
  I18N_LOCAL_USER_LANG,
} from '../resources/constants';

export const initialState = {
  tsl: (term, args) => tsl(en, term, args),
  dictionaries: [en, fr],
  lang: en.LANG,
};

const reducer = (...dictionaries) => {
  let innerState = { ...initialState };

  if (dictionaries && dictionaries[0]) {
    innerState = {
      ...innerState,
      dictionaries,
      lang: dictionaries[0].LANG,
      tsl: (term, args) => tsl(dictionaries[0], term, args),
    };
  }

  return (state = innerState, action = {}) => {
    switch (action.type) {
      case I18N_INIT_LANG:
      case I18N_CHANGE_LANG: {
        const currentIndex = state.dictionaries.findIndex(
          ({ LANG }) => LANG === action.lang
        );

        return {
          ...state,
          ...action,
          tsl: (term, args) =>
            tsl(state.dictionaries[currentIndex], term, args),
        };
      }

      case I18N_SWITCH_LANG: {
        const currentIndex = state.dictionaries.findIndex(
          ({ LANG }) => LANG === state.lang
        );
        const nextIndex = (currentIndex + 1) % state.dictionaries.length;
        const nextDico = state.dictionaries[nextIndex];
        const nextLang = nextDico.LANG;

        localStorage.setItem(I18N_LOCAL_USER_LANG, nextLang);

        return {
          ...state,
          ...action,
          lang: nextLang,
          tsl: (term, args) => tsl(state.dictionaries[nextIndex], term, args),
        };
      }

      default:
        return state;
    }
  };
};

export default reducer;
