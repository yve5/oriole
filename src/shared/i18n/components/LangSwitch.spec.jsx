import React from 'react';
import configureMockStore from 'redux-mock-store';

import { thunk } from 'redux-thunk';
import { Provider } from 'react-redux';
import { act, fireEvent, render } from '@testing-library/react';

import reducer from '../reducer';
import LangSwitch from './LangSwitch';

const mockStore = configureMockStore([thunk]);

const defaultStore = { i18n: reducer()() };

describe('LangSwitch', () => {
  it('should match expected snapshot. default', async () => {
    const { getByText, rerender } = render(
      <Provider store={mockStore(defaultStore)}>
        <LangSwitch />
      </Provider>
    );

    const text = getByText('en');

    await act(async () => {
      fireEvent.click(text);
    });

    expect(text).toBeDefined();
  });
});
