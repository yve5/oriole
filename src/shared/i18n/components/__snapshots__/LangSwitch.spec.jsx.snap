// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LangSwitch should match expected snapshot. default 1`] = `
<button
  data-testid="i18n-switch-button"
  onClick={[Function]}
  type="button"
>
  Français
</button>
`;

exports[`LangSwitch should match expected snapshot. switch lang & fr 1`] = `
<button
  data-testid="i18n-switch-button"
  onClick={[Function]}
  type="button"
>
  Français
</button>
`;

exports[`LangSwitch should match expected snapshot. switch lang 1`] = `
<button
  data-testid="i18n-switch-button"
  onClick={[Function]}
  type="button"
>
  Français
</button>
`;
